File: Src/java/engine-fhir/src/main/java/org/opencds/cqf/cql/engine/fhir/retrieve/BaseFhirQueryGenerator.java
package org.opencds.cqf.cql.engine.fhir.retrieve;

import ca.uhn.fhir.context.FhirContext;
import ca.uhn.fhir.context.FhirVersionEnum;
import ca.uhn.fhir.context.RuntimeSearchParam;
import ca.uhn.fhir.model.api.IQueryParameterType;
import ca.uhn.fhir.rest.api.RestSearchParameterTypeEnum;
import ca.uhn.fhir.rest.param.DateParam;
import ca.uhn.fhir.rest.param.DateRangeParam;
import ca.uhn.fhir.rest.param.ParamPrefixEnum;
import ca.uhn.fhir.rest.param.TokenOrListParam;
import ca.uhn.fhir.rest.param.TokenParam;
import ca.uhn.fhir.rest.param.TokenParamModifier;
import java.util.*;
import org.apache.commons.lang3.tuple.Pair;
import org.hl7.fhir.instance.model.api.IBaseConformance;
import org.hl7.fhir.instance.model.api.ICompositeType;
import org.opencds.cqf.cql.engine.fhir.exception.FhirVersionMisMatchException;
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterMap;
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver;
import org.opencds.cqf.cql.engine.model.ModelResolver;
import org.opencds.cqf.cql.engine.runtime.Code;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Interval;
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo;

public abstract class BaseFhirQueryGenerator implements FhirVersionIntegrityChecker {
    protected static final boolean DEFAULT_SHOULD_EXPAND_VALUESETS = false;

    protected FhirContext fhirContext;

    protected TerminologyProvider terminologyProvider;
    protected SearchParameterResolver searchParameterResolver;
    protected ModelResolver modelResolver;

    private Integer pageSize;

    public Integer getPageSize() {
        return this.pageSize;
    }

    public void setPageSize(Integer value) {
        if (value == null || value < 1) {
            throw new IllegalArgumentException("value must be a non-null integer > 0");
        }
        this.pageSize = value;
    }

    private Integer maxCodesPerQuery;

    public Integer getMaxCodesPerQuery() {
        return this.maxCodesPerQuery;
    }

    public void setMaxCodesPerQuery(Integer value) {
        if (value == null || value < 1) {
            throw new IllegalArgumentException("value must be non-null integer > 0");
        }
        this.maxCodesPerQuery = value;
    }

    private Integer queryBatchThreshold;

    public Integer getQueryBatchThreshold() {
        return this.queryBatchThreshold;
    }

    public void setQueryBatchThreshold(Integer value) {
        if (value == null || value < 1) {
            throw new IllegalArgumentException("value must be non-null integer > 0");
        }
        this.queryBatchThreshold = value;
    }

    // TODO: Think about how to best handle the decision to expand value sets... Should it be part of the
    // terminology provider if it detects support for "code:in"? How does that feed back to the retriever?
    private boolean expandValueSets;

    public boolean isExpandValueSets() {
        return this.expandValueSets;
    }

    public void setExpandValueSets(boolean expandValueSets) {
        this.expandValueSets = expandValueSets;
    }

    protected BaseFhirQueryGenerator(
            SearchParameterResolver searchParameterResolver,
            TerminologyProvider terminologyProvider,
            ModelResolver modelResolver,
            FhirContext fhirContext)
            throws FhirVersionMisMatchException {
        this.searchParameterResolver = searchParameterResolver;
        this.terminologyProvider = terminologyProvider;
        this.modelResolver = modelResolver;
        this.setExpandValueSets(DEFAULT_SHOULD_EXPAND_VALUESETS);

        this.fhirContext = fhirContext;
        validateFhirVersionIntegrity(fetchFhirVersionEnum(this.fhirContext));
    }

    @Override
    public void validateFhirVersionIntegrity(FhirVersionEnum fhirVersionEnum) throws FhirVersionMisMatchException {
        if (this.searchParameterResolver != null
                && fetchFhirVersionEnum(this.searchParameterResolver.getFhirContext()) != fhirVersionEnum) {
            throw new IllegalArgumentException("Components have different");
        }
    }

    public FhirVersionEnum fetchFhirVersionEnum(FhirContext fhirContext) {
        if (fhirContext == null) {
            throw new NullPointerException("The provided argument is null");
        }
        return fhirContext.getVersion().getVersion();
    }

    public abstract List<String> generateFhirQueries(
            ICompositeType dataRequirement,
            DateTime evaluationDateTime,
            Map<String, Object> contextValues,
            Map<String, Object> parameters,
            IBaseConformance capabilityStatement);

    public abstract FhirVersionEnum getFhirVersion();

    protected Pair<String, IQueryParameterType> getTemplateParam(String dataType, String templateId) {
        if (templateId == null || templateId.equals("")) {
            return null;
        }

        // Do something?
        return null;
    }

    protected Pair<String, DateRangeParam> getDateRangeParam(
            String dataType, String datePath, String dateLowPath, String dateHighPath, Interval dateRange) {
        if (dateRange == null) {
            return null;
        }

        DateParam low = null;
        DateParam high = null;
        if (dateRange.getLow() != null) {
            if (dateRange.getLow() instanceof Date) {
                low = new DateParam(
                        ParamPrefixEnum.GREATERTHAN_OR_EQUALS, Date.from(((Date) dateRange.getLow()).toInstant()));
            } else if (dateRange.getLow() instanceof DateTime) {
                low = new DateParam(
                        ParamPrefixEnum.GREATERTHAN_OR_EQUALS,
                        Date.from(((DateTime) dateRange.getLow()).getDateTime().toInstant()));
            }
        }

        if (dateRange.getHigh() != null) {
            if (dateRange.getLow() instanceof Date) {
                high = new DateParam(
                        ParamPrefixEnum.LESSTHAN_OR_EQUALS, Date.from(((Date) dateRange.getHigh()).toInstant()));
            } else if (dateRange.getLow() instanceof DateTime) {
                high = new DateParam(
                        ParamPrefixEnum.LESSTHAN_OR_EQUALS,
                        Date.from(((DateTime) dateRange.getHigh()).getDateTime().toInstant()));
            }
        }

        DateRangeParam rangeParam;
        if (low == null && high != null) {
            rangeParam = new DateRangeParam(high);
        } else if (high == null && low != null) {
            rangeParam = new DateRangeParam(low);
        } else {
            rangeParam = new DateRangeParam(low, high);
        }

        RuntimeSearchParam dateParam = this.searchParameterResolver.getSearchParameterDefinition(
                dataType, datePath, RestSearchParameterTypeEnum.DATE);

        if (dateParam == null) {
            throw new UnsupportedOperationException(String.format(
                    "Could not resolve a search parameter with date type for %s.%s ", dataType, datePath));
        }

        return Pair.of(dateParam.getName(), rangeParam);
    }

    protected Pair<String, IQueryParameterType> getContextParam(
            String dataType, String context, String contextPath, Object contextValue) {
        if (context != null && context.equals("Patient") && contextValue != null && contextPath != null) {
            return this.searchParameterResolver.createSearchParameter(
                    context, dataType, contextPath, (String) contextValue);
        }

        return null;
    }

    protected Pair<String, List<TokenOrListParam>> getCodeParams(
            String dataType, String codePath, Iterable<Code> codes, String valueSet) {
        if (valueSet != null && valueSet.startsWith("urn:oid:")) {
            valueSet = valueSet.replace("urn:oid:", "");
        }

        if (codePath == null && (codes != null || valueSet != null)) {
            throw new IllegalArgumentException("A code path must be provided when filtering on codes or a valueset.");
        }

        if (codePath == null || codePath.isEmpty()) {
            return null;
        }

        // TODO: This assumes the code path will always be a token param.
        List<TokenOrListParam> codeParamLists = this.getCodeParams(codes, valueSet);
        if (codeParamLists == null || codeParamLists.isEmpty()) {
            return null;
        }

        RuntimeSearchParam codeParam = this.searchParameterResolver.getSearchParameterDefinition(
                dataType, codePath, RestSearchParameterTypeEnum.TOKEN);

        if (codeParam == null) {
            return null;
        }

        return Pair.of(codeParam.getName(), codeParamLists);
    }

    // The code params will be either the literal set of codes in the event the data server doesn't have the referenced
    // ValueSet (or doesn't support pulling and caching a ValueSet). If the target server DOES support that then it's
    // "dataType.codePath in ValueSet"
    protected List<TokenOrListParam> getCodeParams(Iterable<Code> codes, String valueSet) {
        if (valueSet != null) {
            if (!isExpandValueSets()) {
                return Collections.singletonList(
                        new TokenOrListParam().addOr(new TokenParam(valueSet).setModifier(TokenParamModifier.IN)));
            } else {
                if (terminologyProvider == null) {
                    throw new IllegalArgumentException(
                            "Expand value sets cannot be used without a terminology provider and no terminology provider is set.");
                }
                ValueSetInfo valueSetInfo = new ValueSetInfo().withId(valueSet);
                codes = terminologyProvider.expand(valueSetInfo);
            }
        }

        if (codes == null) {
            return Collections.emptyList();
        }

        // If the number of queries generated as a result of the code list size and maxCodesPerQuery constraint would
        // result in a number of queries that is greater than the queryBatchThreshold value, then codeParams will
        // be omitted altogether.
        if (getMaxCodesPerQuery() != null && getQueryBatchThreshold() != null) {
            if ((getIterableSize(codes) / (float) getMaxCodesPerQuery()) > getQueryBatchThreshold()) {
                return Collections.emptyList();
            }
        }

        List<TokenOrListParam> codeParamsList = new ArrayList<>();
        TokenOrListParam codeParams = null;
        int codeCount = 0;
        for (Object code : codes) {
            if (getMaxCodesPerQuery() == null) {
                if (codeCount == 0 && codeParams == null) {
                    codeParams = new TokenOrListParam();
                }
            } else if (codeCount % getMaxCodesPerQuery() == 0) {
                if (codeParams != null) {
                    codeParamsList.add(codeParams);
                }

                codeParams = new TokenOrListParam();
            }

            codeCount++;
            if (code instanceof Code) {
                Code c = (Code) code;
                codeParams.addOr(new TokenParam(c.getSystem(), c.getCode()));
            } else if (code instanceof String) {
                String s = (String) code;
                codeParams.addOr(new TokenParam(s));
            }
        }

        if (codeParams != null) {
            codeParamsList.add(codeParams);
        }

        return codeParamsList;
    }

    protected List<SearchParameterMap> setupQueries(
            String context,
            String contextPath,
            Object contextValue,
            String dataType,
            String templateId,
            List<CodeFilter> codeFilters,
            List<DateFilter> dateFilters) {
        Pair<String, IQueryParameterType> templateParam = this.getTemplateParam(dataType, templateId);

        Pair<String, IQueryParameterType> contextParam =
                this.getContextParam(dataType, context, contextPath, contextValue);

        List<Pair<String, DateRangeParam>> dateRangeParams = new ArrayList<Pair<String, DateRangeParam>>();
        if (dateFilters != null) {
            for (DateFilter df : dateFilters) {
                Pair<String, DateRangeParam> dateRangeParam = this.getDateRangeParam(
                        dataType, df.getDatePath(), df.getDateLowPath(), df.getDateHighPath(), df.getDateRange());
                if (dateRangeParam != null) {
                    dateRangeParams.add(dateRangeParam);
                }
            }
        }

        List<Pair<String, List<TokenOrListParam>>> codeParamList =
                new ArrayList<Pair<String, List<TokenOrListParam>>>();
        if (codeFilters != null) {
            for (CodeFilter cf : codeFilters) {
                Pair<String, List<TokenOrListParam>> codeParams =
                        this.getCodeParams(dataType, cf.getCodePath(), cf.getCodes(), cf.getValueSet());

                if (codeParams != null) {
                    codeParamList.add(codeParams);
                }
            }
        }

        return this.innerSetupQueries(templateParam, contextParam, dateRangeParams, codeParamList);
    }

    protected List<SearchParameterMap> setupQueries(
            String context,
            String contextPath,
            Object contextValue,
            String dataType,
            String templateId,
            String codePath,
            Iterable<Code> codes,
            String valueSet,
            String datePath,
            String dateLowPath,
            String dateHighPath,
            Interval dateRange) {
        return setupQueries(
                context,
                contextPath,
                contextValue,
                dataType,
                templateId,
                codePath != null ? Arrays.asList(new CodeFilter(codePath, codes, valueSet)) : null,
                datePath != null || dateLowPath != null || dateHighPath != null
                        ? Arrays.asList(new DateFilter(datePath, dateLowPath, dateHighPath, dateRange))
                        : null);
    }

    protected List<SearchParameterMap> innerSetupQueries(
            Pair<String, IQueryParameterType> templateParam,
            Pair<String, IQueryParameterType> contextParam,
            List<Pair<String, DateRangeParam>> dateRangeParams,
            List<Pair<String, List<TokenOrListParam>>> codeParams) {

        if (codeParams == null || codeParams.isEmpty()) {
            return Collections.singletonList(this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams));
        }

        Pair<String, List<TokenOrListParam>> chunkedCodeParam = null;
        for (Pair<String, List<TokenOrListParam>> cp : codeParams) {
            if (cp.getValue() != null && cp.getValue().size() > 1) {
                if (chunkedCodeParam != null) {
                    throw new IllegalArgumentException(String.format(
                            "Cannot evaluate multiple chunked code filters on %s and %s",
                            chunkedCodeParam.getKey(), cp.getKey()));
                }
                chunkedCodeParam = cp;
            }
        }

        if (chunkedCodeParam == null) {
            return Collections.singletonList(this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams));
        }

        List<SearchParameterMap> maps = new ArrayList<>();
        for (TokenOrListParam tolp : chunkedCodeParam.getValue()) {
            SearchParameterMap base = this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams);
            base.add(chunkedCodeParam.getKey(), tolp);
            maps.add(base);
        }

        return maps;
    }

    protected SearchParameterMap getBaseMap(
            Pair<String, IQueryParameterType> templateParam,
            Pair<String, IQueryParameterType> contextParam,
            List<Pair<String, DateRangeParam>> dateRangeParams,
            List<Pair<String, List<TokenOrListParam>>> codeParams) {
        SearchParameterMap baseMap = new SearchParameterMap();
        baseMap.setLastUpdated(new DateRangeParam());

        if (this.getPageSize() != null) {
            baseMap.setCount(pageSize);
        }

        if (templateParam != null) {
            baseMap.add(templateParam.getKey(), templateParam.getValue());
        }

        if (dateRangeParams != null) {
            for (Pair<String, DateRangeParam> drp : dateRangeParams) {
                baseMap.add(drp.getKey(), drp.getValue());
            }
        }

        if (codeParams != null) {
            for (Pair<String, List<TokenOrListParam>> cp : codeParams) {
                if (cp.getValue() == null
                        || cp.getValue().isEmpty()
                        || cp.getValue().size() > 1) {
                    // NOTE: Ignores "chunked" code parameters so they don't have to be removed
                    continue;
                }
                baseMap.add(cp.getKey(), cp.getValue().get(0));
            }
        }

        if (contextParam != null) {
            baseMap.add(contextParam.getKey(), contextParam.getValue());
        }

        return baseMap;
    }

    private static int getIterableSize(Iterable<?> iterable) {
        if (iterable instanceof Collection) {
            return ((Collection<?>) iterable).size();
        }

        int counter = 0;
        for (Object i : iterable) {
            counter++;
        }
        return counter;
    }
}

File: Src/java/engine-fhir/src/main/kotlin/org/opencds/cqf/cql/engine/fhir/retrieve/BaseFhirQueryGenerator.kt
package org.opencds.cqf.cql.engine.fhir.retrieve

import ca.uhn.fhir.context.FhirContext
import ca.uhn.fhir.context.FhirVersionEnum
import ca.uhn.fhir.model.api.IQueryParameterType
import ca.uhn.fhir.rest.api.RestSearchParameterTypeEnum
import ca.uhn.fhir.rest.param.DateParam
import ca.uhn.fhir.rest.param.DateRangeParam
import ca.uhn.fhir.rest.param.ParamPrefixEnum
import ca.uhn.fhir.rest.param.TokenOrListParam
import ca.uhn.fhir.rest.param.TokenParam
import ca.uhn.fhir.rest.param.TokenParamModifier
import java.util.*
import org.apache.commons.lang3.tuple.Pair
import org.hl7.fhir.instance.model.api.IBaseConformance
import org.hl7.fhir.instance.model.api.ICompositeType
import org.opencds.cqf.cql.engine.fhir.exception.FhirVersionMisMatchException
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterMap
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver
import org.opencds.cqf.cql.engine.model.ModelResolver
import org.opencds.cqf.cql.engine.runtime.Code
import org.opencds.cqf.cql.engine.runtime.DateTime
import org.opencds.cqf.cql.engine.runtime.Interval
import org.opencds.cqf.cql.engine.terminology.TerminologyProvider
import org.opencds.cqf.cql.engine.terminology.ValueSetInfo

abstract class BaseFhirQueryGenerator
protected constructor(
    protected var searchParameterResolver: SearchParameterResolver,
    protected var terminologyProvider: TerminologyProvider?,
    protected var modelResolver: ModelResolver,
    fhirContext: FhirContext,
) : FhirVersionIntegrityChecker {
    protected var fhirContext: FhirContext

    var pageSize: Int? = null
        set(value) {
            require(value != null && value > 0) { "value must be a non-null integer > 0" }
            field = value
        }

    var maxCodesPerQuery: Int? = null
        set(value) {
            require(value != null && value > 0) { "value must be non-null integer > 0" }
            field = value
        }

    var queryBatchThreshold: Int? = null
        set(value) {
            require(value != null && value > 0) { "value must be non-null integer > 0" }
            field = value
        }

    // TODO: Think about how to best handle the decision to expand value sets... Should it be part
    // of the
    // terminology provider if it detects support for "code:in"? How does that feed back to the
    // retriever?
    var isExpandValueSets: Boolean = false

    init {
        this.isExpandValueSets = DEFAULT_SHOULD_EXPAND_VALUESETS

        this.fhirContext = fhirContext
        validateFhirVersionIntegrity(fetchFhirVersionEnum(this.fhirContext))
    }

    @Throws(FhirVersionMisMatchException::class)
    override fun validateFhirVersionIntegrity(fhirVersion: FhirVersionEnum) {
        require(fetchFhirVersionEnum(this.searchParameterResolver.fhirContext) == fhirVersion)
    }

    fun fetchFhirVersionEnum(fhirContext: FhirContext): FhirVersionEnum {
        return fhirContext.version.version
    }

    abstract fun generateFhirQueries(
        dataRequirement: ICompositeType?,
        evaluationDateTime: DateTime?,
        contextValues: MutableMap<String, Any?>?,
        parameters: MutableMap<String, Any?>?,
        capabilityStatement: IBaseConformance?,
    ): MutableList<String>

    protected fun getTemplateParam(
        dataType: String?,
        templateId: String?,
    ): Pair<String, IQueryParameterType>? {
        if (templateId == null || templateId == "") {
            return null
        }

        // Do something?
        return null
    }

    fun getDateRangeParam(
        dataType: String?,
        datePath: String?,
        dateLowPath: String?,
        dateHighPath: String?,
        dateRange: Interval?,
    ): Pair<String, DateRangeParam>? {
        if (dateRange == null) {
            return null
        }

        var low: DateParam? = null
        var high: DateParam? = null
        if (dateRange.low != null) {
            if (dateRange.low is Date) {
                low =
                    DateParam(
                        ParamPrefixEnum.GREATERTHAN_OR_EQUALS,
                        Date.from((dateRange.low as Date).toInstant()),
                    )
            } else if (dateRange.low is DateTime) {
                low =
                    DateParam(
                        ParamPrefixEnum.GREATERTHAN_OR_EQUALS,
                        Date.from((dateRange.low as DateTime).dateTime!!.toInstant()),
                    )
            }
        }

        if (dateRange.high != null) {
            if (dateRange.low is Date) {
                high =
                    DateParam(
                        ParamPrefixEnum.LESSTHAN_OR_EQUALS,
                        Date.from((dateRange.high as Date).toInstant()),
                    )
            } else if (dateRange.low is DateTime) {
                high =
                    DateParam(
                        ParamPrefixEnum.LESSTHAN_OR_EQUALS,
                        Date.from((dateRange.high as DateTime).dateTime!!.toInstant()),
                    )
            }
        }
        val rangeParam =
            if (high == null && low != null) {
                DateRangeParam(low)
            } else {
                DateRangeParam(low, high)
            }

        val dateParam =
            this.searchParameterResolver.getSearchParameterDefinition(
                dataType,
                datePath,
                RestSearchParameterTypeEnum.DATE,
            )

        if (dateParam == null) {
            throw UnsupportedOperationException(
                "Could not resolve a search parameter with date type for $dataType.$datePath "
            )
        }

        return Pair.of<String, DateRangeParam>(dateParam.name, rangeParam)
    }

    protected fun getContextParam(
        dataType: String?,
        context: String?,
        contextPath: String?,
        contextValue: Any?,
    ): Pair<String, IQueryParameterType>? {
        if (
            context != null && context == "Patient" && contextValue != null && contextPath != null
        ) {
            return this.searchParameterResolver.createSearchParameter(
                context,
                dataType,
                contextPath,
                contextValue as String,
            )
        }

        return null
    }

    protected fun getCodeParams(
        dataType: String?,
        codePath: String?,
        codes: Iterable<Code?>?,
        valueSet: String?,
    ): Pair<String, MutableList<TokenOrListParam>>? {
        var valueSet = valueSet
        if (valueSet != null && valueSet.startsWith("urn:oid:")) {
            valueSet = valueSet.replace("urn:oid:", "")
        }

        require(!(codePath == null && (codes != null || valueSet != null))) {
            "A code path must be provided when filtering on codes or a valueset."
        }

        if (codePath == null || codePath.isEmpty()) {
            return null
        }

        // TODO: This assumes the code path will always be a token param.
        val codeParamLists = this.getCodeParams(codes, valueSet)
        if (codeParamLists.isEmpty()) {
            return null
        }

        val codeParam =
            this.searchParameterResolver.getSearchParameterDefinition(
                dataType,
                codePath,
                RestSearchParameterTypeEnum.TOKEN,
            )

        if (codeParam == null) {
            return null
        }

        return Pair.of<String, MutableList<TokenOrListParam>>(codeParam.name, codeParamLists)
    }

    // The code params will be either the literal set of codes in the event the data server doesn't
    // have the referenced
    // ValueSet (or doesn't support pulling and caching a ValueSet). If the target server DOES
    // support that then it's
    // "dataType.codePath in ValueSet"
    protected fun getCodeParams(
        codes: Iterable<Code?>?,
        valueSet: String?,
    ): MutableList<TokenOrListParam> {
        var codes = codes
        if (valueSet != null) {
            if (!this.isExpandValueSets) {
                return mutableListOf(
                    TokenOrListParam()
                        .addOr(TokenParam(valueSet).setModifier(TokenParamModifier.IN))
                )
            } else {
                requireNotNull(terminologyProvider) {
                    "Expand value sets cannot be used without a terminology provider and no terminology provider is set."
                }
                val valueSetInfo = ValueSetInfo().withId(valueSet)
                codes = terminologyProvider!!.expand(valueSetInfo)
            }
        }

        if (codes == null) {
            return mutableListOf()
        }

        // If the number of queries generated as a result of the code list size and maxCodesPerQuery
        // constraint would
        // result in a number of queries that is greater than the queryBatchThreshold value, then
        // codeParams will
        // be omitted altogether.
        if (this.maxCodesPerQuery != null && this.queryBatchThreshold != null) {
            if (
                (getIterableSize(codes) / this.maxCodesPerQuery!!.toFloat()) >
                    this.queryBatchThreshold!!
            ) {
                return mutableListOf()
            }
        }

        val codeParamsList: MutableList<TokenOrListParam> = mutableListOf()
        var codeParams: TokenOrListParam? = null
        var codeCount = 0
        for (code in codes) {
            if (this.maxCodesPerQuery == null) {
                if (codeCount == 0 && codeParams == null) {
                    codeParams = TokenOrListParam()
                }
            } else if (codeCount % this.maxCodesPerQuery!! == 0) {
                if (codeParams != null) {
                    codeParamsList.add(codeParams)
                }

                codeParams = TokenOrListParam()
            }

            codeCount++
            // We have a couple cases where types are erased and code
            // is secretly a String, which the compiler can't detect
            // We need to fix that in the corresponding places.
            @Suppress("USELESS_IS_CHECK")
            if (code is Code) {
                codeParams!!.addOr(TokenParam(code.system, code.code))
            } else if (code is String) {
                codeParams!!.addOr(TokenParam(code))
            }
        }

        if (codeParams != null) {
            codeParamsList.add(codeParams)
        }

        return codeParamsList
    }

    protected fun setupQueries(
        context: String?,
        contextPath: String?,
        contextValue: Any?,
        dataType: String?,
        templateId: String?,
        codeFilters: MutableList<CodeFilter>?,
        dateFilters: MutableList<DateFilter>?,
    ): MutableList<SearchParameterMap> {
        val templateParam = this.getTemplateParam(dataType, templateId)

        val contextParam = this.getContextParam(dataType, context, contextPath, contextValue)

        val dateRangeParams: MutableList<Pair<String, DateRangeParam>> = mutableListOf()
        if (dateFilters != null) {
            for (df in dateFilters) {
                val dateRangeParam =
                    this.getDateRangeParam(
                        dataType,
                        df.datePath,
                        df.dateLowPath,
                        df.dateHighPath,
                        df.dateRange,
                    )
                if (dateRangeParam != null) {
                    dateRangeParams.add(dateRangeParam)
                }
            }
        }

        val codeParamList: MutableList<Pair<String, MutableList<TokenOrListParam>>> =
            ArrayList<Pair<String, MutableList<TokenOrListParam>>>()
        if (codeFilters != null) {
            for (cf in codeFilters) {
                val codeParams = this.getCodeParams(dataType, cf.codePath, cf.codes, cf.valueSet)

                if (codeParams != null) {
                    codeParamList.add(codeParams)
                }
            }
        }

        return this.innerSetupQueries(templateParam, contextParam, dateRangeParams, codeParamList)
    }

    fun setupQueries(
        context: String?,
        contextPath: String?,
        contextValue: Any?,
        dataType: String?,
        templateId: String?,
        codePath: String?,
        codes: Iterable<Code?>?,
        valueSet: String?,
        datePath: String?,
        dateLowPath: String?,
        dateHighPath: String?,
        dateRange: Interval?,
    ): MutableList<SearchParameterMap> {
        return setupQueries(
            context,
            contextPath,
            contextValue,
            dataType,
            templateId,
            if (codePath != null) mutableListOf(CodeFilter(codePath, codes, valueSet)) else null,
            if (datePath != null || dateLowPath != null || dateHighPath != null)
                mutableListOf(DateFilter(datePath, dateLowPath, dateHighPath, dateRange))
            else null,
        )
    }

    protected fun innerSetupQueries(
        templateParam: Pair<String, IQueryParameterType>?,
        contextParam: Pair<String, IQueryParameterType>?,
        dateRangeParams: MutableList<Pair<String, DateRangeParam>>?,
        codeParams: MutableList<Pair<String, MutableList<TokenOrListParam>>>?,
    ): MutableList<SearchParameterMap> {
        if (codeParams == null || codeParams.isEmpty()) {
            return mutableListOf<SearchParameterMap>(
                this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams)
            )
        }

        var chunkedCodeParam: Pair<String, MutableList<TokenOrListParam>>? = null
        for (cp in codeParams) {
            if (cp.value != null && cp.value!!.size > 1) {
                require(chunkedCodeParam == null) {
                    "Cannot evaluate multiple chunked code filters on ${chunkedCodeParam!!.key} and ${cp.key}"
                }
                chunkedCodeParam = cp
            }
        }

        if (chunkedCodeParam == null) {
            return mutableListOf(
                this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams)
            )
        }

        val maps: MutableList<SearchParameterMap> = mutableListOf()
        for (t in chunkedCodeParam.value!!) {
            val base = this.getBaseMap(templateParam, contextParam, dateRangeParams, codeParams)
            base.add(chunkedCodeParam.key!!, t)
            maps.add(base)
        }

        return maps
    }

    fun getBaseMap(
        templateParam: Pair<String, IQueryParameterType>?,
        contextParam: Pair<String, IQueryParameterType>?,
        dateRangeParams: MutableList<Pair<String, DateRangeParam>>?,
        codeParams: MutableList<Pair<String, MutableList<TokenOrListParam>>>?,
    ): SearchParameterMap {
        val baseMap = SearchParameterMap()
        baseMap.lastUpdated = DateRangeParam()

        if (this.pageSize != null) {
            baseMap.count = pageSize
        }

        if (templateParam != null) {
            baseMap.add(templateParam.key!!, templateParam.value)
        }

        if (dateRangeParams != null) {
            for (drp in dateRangeParams) {
                baseMap.add(drp.key!!, drp.value)
            }
        }

        if (codeParams != null) {
            for (cp in codeParams) {
                if (cp.value == null || cp.value!!.isEmpty() || cp.value!!.size > 1) {
                    // NOTE: Ignores "chunked" code parameters so they don't have to be removed
                    continue
                }
                baseMap.add(cp.key!!, cp.value!![0])
            }
        }

        if (contextParam != null) {
            baseMap.add(contextParam.key!!, contextParam.value)
        }

        return baseMap
    }

    companion object {
        protected const val DEFAULT_SHOULD_EXPAND_VALUESETS: Boolean = false

        private fun getIterableSize(iterable: Iterable<*>): Int {
            if (iterable is MutableCollection<*>) {
                return iterable.size
            }

            var counter = 0
            for (i in iterable) {
                counter++
            }
            return counter
        }
    }
}
